validator/client/validator.go

```diff
@@ -130,6 +130,89 @@ func (v *validator) Done() {
 	v.ticker.Done()
 }
 
+func (v *validator) Init(ctx context.Context) error {
+	ctx, span := trace.StartSpan(ctx, "validator.Init")
+	defer span.End()
+
+	ticker := time.NewTicker(backOffPeriod)
+	defer ticker.Stop()
+
+	firstTime := true
+	var (
+		currentSlot primitives.Slot
+		err         error
+	)
+	for {
+		if !firstTime {
+			if ctx.Err() != nil {
+				log.Info("Context canceled, stopping validator")
+				return errors.New("context canceled")
+			}
+			<-ticker.C
+		}
+
+		firstTime = false
+		currentSlot, err = v.WaitForChainStart(ctx)
+		if err != nil {
+			if isConnectionError(err) {
+				log.WithError(err).Warn("Could not determine if beacon chain started")
+				continue
+			}
+
+			return errors.Wrap(err, "Could not determine if beacon chain started")
+		}
+
+		if err := v.WaitForKeymanagerInitialization(ctx); err != nil {
+			// log.Fatal will prevent defer from being called
+			v.Done()
+			return errors.Wrap(err, "Wallet is not ready")
+		}
+
+		if err := v.WaitForSync(ctx); err != nil {
+			if isConnectionError(err) {
+				log.WithError(err).Warn("Could not determine if beacon chain started")
+				continue
+			}
+
+			return errors.Wrap(err, "Could not determine if beacon node synced")
+		}
+
+		if err := v.WaitForActivation(ctx, nil /* accountsChangedChan */); err != nil {
+			return errors.Wrap(err, "Could not wait for validator activation")
+		}
+
+		if err := v.CheckDoppelGanger(ctx); err != nil {
+			if isConnectionError(err) {
+				log.WithError(err).Warn("Could not wait for checking doppelganger")
+				continue
+			}
+
+			return errors.Wrap(err, "Could not succeed with doppelganger check")
+		}
+		break
+	}
+
+	// should there be a check if it's too later into current slot?
+	if err := v.UpdateDuties(ctx, currentSlot); err != nil {
```

Comment:
we need this UpdatDuties in case it's the genesis slot, in which case the main run loop would not run on that, need to test on kurtosis

Reply:
Can you explain in more detail why this is needed for genesis slot? Also there is a bug here. `currentSlot` at the time when `WaitForChainStart` can be an entirely different slot than when you call `UpdateDuties`. The most likely reason is that you have to wait for sync. So unless your node is already synced, it will return the wrong slot. You should ask for the current slot right before getting duties. I think `slots.CurrentSlot(chainStartRes.GenesisTime)` can be used if you return the genesis time from `WaitForChainStart`.

Reply:
oh thanks yeah that's a good call out yeah i see the issue now thanks. will retest and check logs carefully.

the main loop will always skip the first slot so we need it for that, we also do it so we don't need to wait the full 12 seconds to get duties again